src/itwinai/torch/trainer.py

```diff
@@ -1212,42 +1180,420 @@ def distributed(func):
+        self._set_scaling_config()
+        self._set_tune_config()
+        self._set_run_config()
+        self._set_train_loop_config()
+
+    @property
+    def device(self) -> str:
+        """Get the current device from distributed strategy.
+        Returns:
+            str: Device string (e.g., "cuda:0").
+        """
+        return self.strategy.device()
+
+    def create_dataloaders(
+        self,
+        train_dataset: Dataset,
+        validation_dataset: Dataset | None = None,
+        test_dataset: Dataset | None = None,
+        batch_size: int = 1,
+        shuffle_train: bool | None = False,
+        shuffle_test: bool | None = False,
+        shuffle_validation: bool | None = False,
+        sampler: Union[Sampler, Iterable, None] = None,
+        collate_fn: Callable[[List], Any] | None = None,
+    ) -> None:
+        """Create data loaders for training, validation, and testing.
+
+        Args:
+            train_dataset (Dataset): The training dataset.
+            validation_dataset (Dataset, optional): The validation dataset. Defaults to None.
+            test_dataset (Dataset, optional): The test dataset. Defaults to None.
+            batch_size (int, optional): Batch size for data loaders. Defaults to 1.
+            shuffle_train (bool, optional): Whether to shuffle the training dataset.
+                Defaults to False.
+            shuffle_test (bool, optional): Whether to shuffle the test dataset.
+                Defaults to False.
+            shuffle_validation (bool, optional): Whether to shuffle the validation dataset.
+                Defaults to False.
+            sampler (Union[Sampler, Iterable, None], optional): Sampler for the datasets.
+                Defaults to None.
+            collate_fn (Callable[[List], Any], optional):
+                Function to collate data samples into batches. Defaults to None.
+        """
+        self.train_dataloader = self.strategy.create_dataloader(
+            dataset=train_dataset,
+            batch_size=batch_size,
+            shuffle=shuffle_train,
+            sampler=sampler,
+            collate_fn=collate_fn,
+        )
+        if validation_dataset is not None:
+            self.validation_dataloader = self.strategy.create_dataloader(
+                dataset=validation_dataset,
+                batch_size=batch_size,
+                shuffle=shuffle_validation,
+                sampler=sampler,
+                collate_fn=collate_fn,
+            )
+        else:
+            self.validation_dataloader = None
+        if test_dataset is not None:
+            self.test_dataloader = self.strategy.create_dataloader(
+                dataset=test_dataset,
+                batch_size=batch_size,
+                shuffle=shuffle_test,
+                sampler=sampler,
+                collate_fn=collate_fn,
+            )
+        else:
+            self.test_dataloader = None
+
+    @abstractmethod
+    def train(config, data=None):
+        pass
+
```

Comment:
Okay this is a good place to discuss moving the inheritance from the `Trainer` to the `TorchTrainer`. I definitely agree that the duplication of many methods, such as `create_dataloaders` or `set_epoch`  are not great. On the other hand, most methods in the `TorchTrainer` do not quite work with the `RayTorchTrainer`, mostly for the RayDDPStrategy, which is the default strategy provided by ray for training. For example, the moving of data to devices is done automatically when creating the DataLoader, and you can't call `enumerate(self.dataloader)`, that will throw an error. So most of the functions that are specific to training, such as `train`, `train_epoch`, etc, would have to be overwritten. For now I would prefer to leave them empty, so I could do something like the above 
```
@abstractmethod
    def train(config, data=None):
        pass
```
for all of the functions that would not work as-is with the `RayTorchTrainer`. 

Do you have a preference, as in having the RayTorchTrainer inherit from the TorchTrainer with this in mind, or leaving it the way it is?

Reply:
I see. Forcing some methods to be abstract just to avoid inheriting them is not a good idea, because the class cannot be instantiated without overriding *all* the abstract methods, even if unused. Perhaps a better idea would be moving the methods which are in common between RayTorchTrainer and TorchTrainer in a common parent class, e.g. BaseTorchTrainer, form which both trainers inherit